### DISCUSSION TITLE
JCB JUG - Meeting Notes - 2022-04-04

### DISCUSSION THREAD

**Discussion by cpaschen** [joomengine/Joomla-Component-Builder#888] [General]
Here are some notes from today's meeting.
Hopefully these provide a quick update to anyone who couldn't be there.
Main 'topic' today was the status of J4 along with a very detailed explanation of what needs to happen to get to J4 and the establishment of a Working Group to help move the process ahead.
_(feel free to update/append as needed)_
[JCB JUG 2022-04-04-.md](https://github.com/vdm-io/Joomla-Component-Builder/files/8412370/JCB.JUG.2022-04-04-.md)
[JCB JUG 2022-04-04.pdf](https://github.com/vdm-io/Joomla-Component-Builder/files/8412371/JCB.JUG.2022-04-04.pdf)

# Group 'business'

- Meeting - 1st Monday of each Month
- We will not record the meetings
    - To motivate people to attend live
    - To allow participants to speak more 'freely' (about projects, etc.) without their comments being shared publicly in a video

# J4

- New Release (just released) 
    - JCB Powers Library
        - Adding Utility classes
            - Some of these classes used to (and still) live in the helper classes
    - Moving into a namespace implementation
        - ex. Array helper has an array helper class
    - "Add new utilities abstract class" addition in the git
        - Adding namespacing
        - Add changes to the compiler
- Everything that we do to move JCB to J4, it will automatically make all the extensions built with it do the same thing
- If you are using the old classes  (CompontentbuilderHelper::checkArray)
    you will need to change to the new (ArrayHelper::check)
- JED
    - Working on JED checker to alert you to things that need to be changed
- Have been changing JCB to move to the namespacing structure of J4
    - Problem during this process ...
        - JCB Pro - has a new area called powers (power/powers)
        - All the extra classes are build in the powers area
        - In the Utilities class - adding all the helper class & creating wrappers
            - Adds the Utilities as a 'trait'
        - To make this available to everyone Powers will need to be public (or everyone will need to buy in - not just for pro)
        - This will make the move to J4 better for all
- Working group for J4 development
    - All members would have access to a joint system
    - Group will meet more often (possibly 1x week)
    - More collaborative distribution method of the Powers code
- Snippets
    - Has get snippets - gets snippets from github
    - repository on github - where people would contribute snippets
    - Eventually hoping that this will take off
    - "Jooma-component builder snippets"
        - See example snippet
    - In site view in the 'select a snippet' you can use the snippet in your code (copy paste)
        - This is managed by that snippets repository
- Moving everything how JCB builds models, controllers, etc. ... into this type of approach
    - Central repository where they are maintained
    - Within powers you can control what is done, how things are done
        - ex. Have different way to build a certain type of component, with a power that can control that and be shared
    - Similar to creating dynamic emails
        - using GUI inserts like you would use fields in a dynamic email
    - Moving the compiler to powers will be the basis of moving to J4
- What can we do to help?
    - Become part of the volunteer group - working group
    - Access that central J site
- Time Frame
    - Wanted to have it done 3 months ago
    - Every time we give time frame we miss it, so reluctant to give it a time frame
    - Was going to do a 'legacy' approach (create a J3 component that would work in J4)
    - Decided no use creating legacy Joomla 4 code
    - The new J4 architecture does things better
    - We are doing fairly complex things in JCB now to make it work, we won't need to to do that in native J4 - so we need to 'dumb down' JCB.
    - Not an easy process to refactor the compiler
        - we will be modernizing the way the compiler works
    - Already done
        - compiler/joomla3/ settings.json
            - settings used to build your whole system's file structure
            - Detect the files to use
            - Global settings for building everything
            - you could override the settings.json and change the way it compiles
            - if you change that file it changes for all components
            - Added the ability to have a settings file for specific components
            - Allows you to make custom changes to files
                - updates do not replace your changes
    - Hesitant to set a date in June at this point
    - Some developers have reached out to see if we could put funding on the table to get it going faster
        - Not eager to do that model
        - Add new sponsoring tab on sponsor tab - specifically for J4 transition
            (on the Open Collective page)
        - If we get enough sponsorship it will speed up the process
    - Not a question of ARE we going to do it but WHEN
    - No way for us to move to J5 without going to J4
    - Compiler is stable but is designed by one
        - Ideally would have multiple contributors and has other input from other more-advanced
- Moving the compiler to a more abstract structure
    - More interest by other/outside developers
    - Could bring more life to Joomla
    - Will make it even more powerful to build many different possibilities than current compiler
    - By making things module - the compiler can target different outputs/structures
    - Will it cause a slowing of the compiler
        - Some others don't feel that it will be an issue

# Other topics

## Backward compatibility of code

- - C++ still respects 'old people' - you can still write C in C++
        - PHP doesn't necessarily respect 'old people' 
            - things break with new code
    - JCB still has code to support older extensions that JCB built
        - you can still build it
    - Learning Management (public) is the oldest extension
    - JCB is the 2nd
    - Stable because it is being used every day
    - Used by the developer/s by own projects

## Working Group

- - - More meetings shorter discussions
        - Possibly via CodeSprints
        - Meet Monday evenings at 8
            - 1st Monday will still be the regular JUG
            - Other Mondays will be the working group meeting
            - Starting next Monday
            - Notice will be posted in the JCB Telegram group
            - Live meeting will be held there
        - Telegram will be the communication channel for the group
        - Llewellyn to do most coding to start but to include others
            - collaborative review of how things are being done
        - Files will be stored on JCB git (not github)
        - If we make changes, how will that look to 'older people' (older projects)
        - Specific development discussions:
            - Concern about speed of compiler
            - Make sure that the compile process doesn't crash (or require extra system resources/memory)
        - Watch YouTube video "JCB Compiler Explained (the code)" playlist
            - Write down any questions that you have after watching
            - Bring questions to the meeting
        - Planning to participate: Llewellyn, Martin, Georgios Papadakis (FLEXIcontent), Emmanuel

**Comment by TLWebdesign** [comment]
Wow very detailed. Thanks! I'm impressed. 😮
